### PR TITLE
Add quantized sigmoid, fix quantized tanh for QKeras

### PR DESCRIPTION
**Pull request by jmitrevs**
This is an attempt to fix quantized tanh and add quantized sigmoid activation support for QKeras ingestion. In Quartus the tanh handling probably needs an optimizer to rename it to dense_tanh (and that is why the tests fail--the hacky fix in the writer doesn't work), but for Vivado I think this generally works. Overall, though, I think this fix is only temporary. The method breaks if `granularity='name'` is not used in `hls4ml.utils.config_from_keras_model`, though this feature is not just for these quantizers.

Any suggestions?

### CODE REVIEW

File: test/pytest/test_qkeras.py
Comment (thesps) [inline test/pytest/test_qkeras.py:267]: Why compare to a floating point version of the function on not the quantized one here?
i.e. you could do `y_calc = quantizer(X)`.
And then, on a related note, is there a problem just including these quantizers in `test_quantizer`?

Comment (jmitrevs) [reply test/pytest/test_qkeras.py:267]: The reason for the different test is to do the different comparison. You have to do a try catch on the integer field, but in general, even if doing `use_qkeras_tanh=True` there were occasional differences that I could not solve by setting rounding and saturation modes. I can look more.

**Comment by jmitrevs** [comment]
I agree about moving the logic to an optimizer pass, but that would be for all the quantizers, not just for quantized tanh and sigmoid. Basically all QKeras info is lost if `granularity='name'` is not used, so it is a more fundamental change we should discuss.

File: test/pytest/test_qkeras.py
Comment (thesps) [reply test/pytest/test_qkeras.py:267]: I think it's quite likely that the `sigmoid` and `tanh` C++ needs a refresh to get bit-exactness wrt QKeras, they both have some hard coded values still (the `init_{sigmoid | tanh}_table` functions in particular)

File: hls4ml/utils/config.py
Comment (thesps) [inline hls4ml/utils/config.py:51]: Maybe move these into the `if quantizer['class_name'] in supported_quantizers:` statement, as an `elif` after `if quantizer['class_name'] == 'quantized_relu'` - just to improve readability so all the special cases after the logic that applies to them all.

File: hls4ml/utils/config.py
Comment (thesps) [inline hls4ml/utils/config.py:42]: These will apply to weights as well as activations. It may not matter, since the 'rounding' for weights actually happens in Python and not HLS, but we should make sure it doesn't have any implications on the HLS results.

**Comment by thesps** [comment]
Also can you push to a local branch (like `pr/569`) to trigger Gitlab CI for pytests?

**Comment by jmitrevs** [comment]
Let me do a few more tests first, then I'll push to a local branch.

File: hls4ml/utils/config.py
Comment (jmitrevs) [reply hls4ml/utils/config.py:42]: I was wondering whether I should put the ap_fixed defaults (AP_TRN and AP_WRAP) or these as the default, and didn't really have convincing arguments one way or the other. The QONNX route tends to have saturation and rounding, so I chose that at the end, but it was not a strong argument.

Comment (thesps) [reply hls4ml/utils/config.py:42]: I think we just need to see that:
- no QKeras tests that were passing now fail
- resources are the same for some synthesized QKeras models on master and on this branch, with all else being the same

**Comment by jmitrevs** [comment]
The issue that we are seeing is that the hls4ml sigmoid really seems to approximate `1/1+exp(-x)`, but the QKeras one is quite different in a way I don't currently understand.

**Comment by jmitrevs** [comment]
So qkeras uses the hard sigmoid by default:
https://github.com/google/qkeras/blob/master/qkeras/quantizers.py#L192-L237
I have to see if hls4ml implements it; otherwise, it's easy to do:  `clip(0.5 * x + 0.5, 0.0, 1.0)`.

File: hls4ml/utils/config.py
Comment (jmitrevs) [reply hls4ml/utils/config.py]: Done

**Comment by jmitrevs** [comment]
This resulted in more changes than I expected. Please take a look to see if this is reasonable.

**Comment by jmitrevs** [comment]
I am not sure how the matplotlib import came into this pull request. It should not be there. I will fix that.

**Comment by jmitrevs** [comment]
Merged in master to fix pytest setup.

File: test/pytest/test_qkeras.py
Comment (jmitrevs) [reply test/pytest/test_qkeras.py]: This has changed so I will resolve this conversion.

File: hls4ml/utils/config.py
Comment (jmitrevs) [reply hls4ml/utils/config.py:42]: I was wondering if we should defer this to the OutputRoundingSaturationMode optimizer. That might be more consistent with what is done in other cases. I will try that.

Comment (jmitrevs) [reply hls4ml/utils/config.py]: I decided to go back to using the ap_fixed defaults.

**Comment by jmitrevs** [comment]
I pushed the changes to pr/569

**Comment by thesps** [comment]
I notice in the qkeras pytest we get:
`E   TypeError: __init__() got an unexpected keyword argument 'use_real_tanh'`
It's because that option is in QKeras main branch but not in a release.

File: hls4ml/converters/keras/qkeras_layers.py
Comment (thesps) [inline hls4ml/converters/keras/qkeras_layers.py:94]: Since these options are not in a QKeras release yet, maybe we should access them with a `.get` returning a default value of `False` for robustness

Comment (jmitrevs) [reply hls4ml/converters/keras/qkeras_layers.py]: I just committed the fix.

**Comment by thesps** [comment]
So a few tests are still failing, including the existing `hard_sigmoid` tests in `test_activations`, some of which fail and at least one of which now creates a segmentation fault. The test with these settings produces the segfault:

```
backend, shape, io_type, activation, name = 'Quartus', (8, ), 'io_stream', Activation('hard_sigmoid'), 'hard_sigmoid'
```

Apart from that, I think we should remove the CNN fixes to a separate PR to keep this one on topic.

**Comment by jmitrevs** [comment]
The CNN fixes are actually in #610. You are right that it should not be here. I am not sure if it's best to try to get that in first or to revert the commit.

**Comment by jmitrevs** [comment]
I fixed up #610 so maybe it's easier to merge this after that PR (and I fix the remaining issues).

**Comment by jmitrevs** [comment]
I think this is ready, but I would like to stage it after #610 (which is also ready). That way the encoded fix, that's partially included here, does not need to be removed. After #610 is merged, I would need to glance one more time at this PR.

File: hls4ml/converters/keras_to_hls.py
Comment (jmitrevs) [inline hls4ml/converters/keras_to_hls.py:269]: Double-check if QActivation should be here.

**Comment by jmitrevs** [comment]
It still passes with those values. I do wonder if it's the qkeras version. I noticed after doing git pull on the master branch the pytests no longer work for unrelated reasons (`TypeError: Got an unexpected keyword argument 'use_legacy_format'`), and checkout out v0.9.0 fails for compatibility reasons. I probably need to target v0.9.0 and make sure that the tests succeed with that version (and disable quantized sigmoid, which is not supported in that version.)

**Comment by vloncar** [comment]
`use_legacy_format` comes from keras serialization functions. You need the latest TF installation for QKeras main. That's a bummer, we need them to make a release and fix on that. 0.9.0 is getting a bit old.

**Comment by jmitrevs** [comment]
So the test fails with tensorflow 2.4.1 (used by the CI) but passes with 2.9.1 and 2.11.0. I don't know why. Should I increase the allowed difference?

**Comment by jmitrevs** [comment]
The pytests now succeed. I am not sure why there are the CI failures, though. Is it because we couldn't build the docker image with Vivado?

**Comment by vloncar** [comment]
Looks like a real error, probably something was generated with a syntax error.

**Comment by jmitrevs** [comment]
I was able to manually reproduce it. It is in the vivado_accelerator_writer not liking the pre-commit changes:
```python
            elif 'void myproject(' in line:
                newline = f'void {model.config.get_project_name()}_axi(\n'
```
This basically removes that arguments that were added to the same line by the pre-commit:
```C++
void myproject(input_axi_t in[N_IN], output_axi_t out[N_OUT]);
```
The fix should be easy. I'll try to fix it also in vivado_writer where it currently doesn't cause a problem but could.

**Review by vloncar — APPROVED** [review]

**Pull request merged**
